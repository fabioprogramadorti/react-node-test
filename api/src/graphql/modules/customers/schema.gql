type Customer {
	_id: ID!
	first_name: String!
	last_name: String!
	email: String!
	gender: String!
	company: String!
	city: String!
	title: String!
	lat: String
	long: String
}

type Total {
	city: String!
	customers_total: Int!
}

type Query {
	totalCustomersByCity: [Total]!
	customersByCity(city: String!):[Customer]!
	customer(id: ID!): Customer!
}

input CustomerInput {
	first_name: String!
	last_name: String!
	email: String!
	gender: String!
	company: String!
	city: String!
	title: String!
}

type Mutation {
	createCustomer(data: CustomerInput!): Customer!
	updateCustomer(id: ID!, data: CustomerInput!): Customer!
	deleteCustomer(id: ID!): Boolean!
}