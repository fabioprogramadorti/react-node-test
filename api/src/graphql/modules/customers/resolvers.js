import Customer from '../../../models/customer'
import getLocationByCity from './googleMaps.service'
import paginatedResults from './utils'

export default {
	Query: {
		totalCustomersByCity: async (_, args) => {

			// group customers by city and count them
			const allCustomers = await Customer.aggregate([
				{
					$group: {
						_id: '$city',
						customers_total: { $sum: 1 }
					}
				}
			])

			// change object key _id by city
			return allCustomers.map(customer => {
				customer.city = customer._id
				delete customer._id
				return customer
			})
		},
		customersByCity: async(_, {city, pageSize=20, after}) => {
			const customersByCity = await Customer.find({city: city})
			const results = customersByCity.map(customer => {
				customer._id = customer._id.toString()
				return customer
			})
			const customers = paginatedResults({
				after,
				pageSize,
				results
			})
			
			return {
				customers,
				cursor: customers.length ? customers[customers.length - 1].id : null,
				hasMore: customers.length 
				? customers[customers.length - 1].id !== 
					customersByCity[customersByCity.length - 1].id
				: false
			}
		},
		customer: async (_, { id }) => {
			let customer = await Customer.findById(id)
			var loc = await getLocationByCity(customer.city)

			customer.lat = loc.lat
			customer.long = loc.lng

			return customer
		}
	}
}