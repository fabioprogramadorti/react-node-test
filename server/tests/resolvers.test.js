import { makeExecutableSchema } from '@graphql-tools/schema';
import { addMocksToSchema } from '@graphql-tools/mock';
import { graphql } from 'graphql';
import typeDefs from '../src/graphql/typeDefs'
import resolvers from '../src/graphql/resolvers'

let schema = makeExecutableSchema({ typeDefs, resolvers })

// Create a new schema with mocks
let schemaWithMocks = addMocksToSchema({ schema });


it("Get number of customers by city", async () => {
	const GET_TOTAL_CUSTOMERS_BY_CITY = `
		query {
				totalCustomersByCity{
					customers_total
					city
				}
			}
		`
	const { data } = await graphql(schemaWithMocks, GET_TOTAL_CUSTOMERS_BY_CITY)

	const { totalCustomersByCity } = data;

	expect(totalCustomersByCity).toMatchSnapshot();

});

it("Get customers by city", async () => {
	const GET_CUSTOMERS_BY_CITY = `
		query customersByCity{
			customersByCity(city:"Some City"){
				cursor
				hasMore
				customers{
					_id
					first_name
				}
			} 
		}
		`
	const { data } = await graphql(schemaWithMocks, GET_CUSTOMERS_BY_CITY)
	const { customersByCity } = data

	expect(customersByCity).toMatchSnapshot();

})

it("Get customer by ID", async () => {
	const GET_CUSTOMER = `
		query customer{
			customer(id: "1234"){
				_id
				first_name
				last_name
				email
				gender
				company
				city
				title
				lat
				long
			}
		}
		`

	const { data } = await graphql(schemaWithMocks, GET_CUSTOMER)
	const { customer } = data
	
	expect(customer).toMatchSnapshot();

});




