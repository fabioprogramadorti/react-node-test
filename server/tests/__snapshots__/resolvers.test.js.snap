// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get customer by ID 1`] = `
Object {
  "_id": "fa9cea31-5df7-4224-af11-32ffe8390714",
  "city": "Hello World",
  "company": "Hello World",
  "email": "Hello World",
  "first_name": "Hello World",
  "gender": "Hello World",
  "last_name": "Hello World",
  "lat": -99.91888169282296,
  "long": 48.90336925743779,
  "title": "Hello World",
}
`;

exports[`Get customers by city 1`] = `
Object {
  "cursor": "Hello World",
  "customers": Array [
    Object {
      "_id": "4de5ad7e-fc6e-464f-8e2d-54d2ccd619b0",
      "first_name": "Hello World",
    },
    Object {
      "_id": "0fe5c2ba-4401-49c7-80fb-2a657678bf94",
      "first_name": "Hello World",
    },
  ],
  "hasMore": false,
}
`;

exports[`Get number of customers by city 1`] = `
Array [
  Object {
    "city": "Hello World",
    "customers_total": 74,
  },
  Object {
    "city": "Hello World",
    "customers_total": -79,
  },
]
`;
